### PR TITLE
implemented warm start

### LINKED ISSUE #35: Warm restart
## Current situation
Currently, models are either stored after each epoch (large memory footprint) or at the end of the training (warm restart not possible).

## Storing approach
To retrain or fine-tune a model, we need to store the model and optimizer state of the last successful epoch. With every new successful epoch, we delete the previous state.  

## Warm restart
The following changes need to made:
1. pass retrain flag to the starter script
2. trainer and evaluator need to be set to correct epoch
3. trainer needs to load the last training state

### PR DESCRIPTION
**Pull request by le1nux**
Closes #35

### CODE REVIEW

**Comment by le1nux** [comment]
There were some missing changes in the latest dashifyML version. It should be fixed now. Would you mind retrying please?

**Comment by PriyaTomar** [comment]
**Problem:** There is an error in calculating the training epochs count for the trainer. Currently in _jobs.py,_ _execute_train()_ has the trainer epoch assignments like following :

`def _execute_train(self, device: torch.device):
        trained_epochs = max(DashifyReader.get_last_epoch(self.experiment_info), 0)
        **self.trainer.set_num_epochs(num_epochs=self.epochs)**
        self.trainer.set_current_epoch(trained_epochs+1)
        while not self.trainer.is_done():
            current_epoch = self.trainer.current_epoch
            # self.logger.log(LogLevel.DEBUG, "Executing evaluation step")
            self._evaluation_step(device, current_epoch-1)
            # self.logger.log(LogLevel.DEBUG, "Executing training step")
            model = self._train_step(device, epoch=current_epoch)
        self._evaluation_step(device, current_epoch)`

Trainer epoch counts does not include the count of previously trained epochs. Hence the while loop is not executed because the loop condition is not evaluated True. As a result, error occurs in using _current_epoch_ after the loop.

The error log is:
`File "/cluster/priya/mlgym/src/ml_gym/multiprocessing/worker.py", line 35, in _do_calc
    job.execute()
  File "/cluster/priya/mlgym/src/ml_gym/multiprocessing/job.py", line 36, in execute
    return self.fun(**self.param_dict)
  File "/cluster/priya/mlgym/src/ml_gym/gym/gym.py", line 46, in _run_job
    decorated_runner(device=device)
  File "/opt/conda/envs/mlgym_env/lib/python3.9/site-packages/dashify/logging/dashify_logging.py", line 248, in decorate_run
    self.run_fun_with_reraise(run_fun, fun_params, file=None)
  File "/opt/conda/envs/mlgym_env/lib/python3.9/site-packages/dashify/logging/dashify_logging.py", line 268, in run_fun_with_reraise
    raise e
  File "/opt/conda/envs/mlgym_env/lib/python3.9/site-packages/dashify/logging/dashify_logging.py", line 264, in run_fun_with_reraise
    run_fun(**fun_params)  # here we call the scripts run method
  File "/cluster/priya/mlgym/src/ml_gym/gym/jobs.py", line 158, in execute
    self._execution_method(device)
  File "/cluster/priya/mlgym/src/ml_gym/gym/jobs.py", line 170, in _execute_train
    self._evaluation_step(device, current_epoch)
UnboundLocalError: local variable 'current_epoch' referenced before assignment`

**Solution**: self.trainer.set_num_epochs(num_epochs= trained_epochs + self.epochs)

**Comment by le1nux** [comment]
oh yeah, I think you tested it without the `--keep-interim-results` flag and you were thus using `GymJobLite` which only stores the model after  the *final* epoch. So far the warm start only works when including `--keep-interim-results` / using `GymJob`.  

I think we should should deprecate GymJobLite and look for a generic model saving approach within `GymJob`, as suggested in issue #36.

**Comment by PriyaTomar** [comment]
Yes I did not set the flag _keep_interim_results_ . I could train a model second time with WARM_START when I do  _self.trainer.set_num_epochs(num_epochs= trained_epochs + self.epochs)_ . 
However, in both the approaches, it throws the error following error if we want to train again for 3rd time .

` File "/opt/conda/envs/mlgym_env/lib/python3.9/site-packages/dashify/logging/dashify_logging.py", line 264, in run_fun_with_reraise
    run_fun(**fun_params)  # here we call the scripts run method
  File "/cluster/priya/mlgym/src/ml_gym/gym/jobs.py", line 158, in execute
    self._execution_method(device)
  File "/cluster/priya/mlgym/src/ml_gym/gym/jobs.py", line 161, in _execute_train
    trained_epochs = max(DashifyReader.get_last_epoch(self.experiment_info), 0)
  File "/cluster/priya/mlgym/src/ml_gym/persistency/io.py", line 123, in get_last_epoch
    raise TrainingStateCorruptError(f"The experiment {experiment_info.experiment_id} has checkpointed the model at epoch {model_epoch} and but has evaluated the model until epoch {metric_epoch}.")
ml_gym.error_handling.exception.TrainingStateCorruptError: The experiment 2022-04-26--19-07-38/conv_net/0 has checkpointed the model at epoch 8 and but has evaluated the model until epoch 9.`

In short, with _WARM_START_, model is evaluated one more time than the trained epochs.
